### DISCUSSION TITLE
How to duplicate page with AcroForm fields and correctly rename fields on the new (copied) page

### DISCUSSION THREAD

**Discussion by ekzobrain** [gettalong/hexapdf#250] [Q&A] [answered]
[page duplicate.pdf](https://github.com/gettalong/hexapdf/files/11886698/page.duplicate.pdf)
[page duplicate result 1.pdf](https://github.com/gettalong/hexapdf/files/11886699/page.duplicate.result.1.pdf)

Hi @gettalong .

I need to copy an existing PDF page with AcroForm fields and then rename fields on the copied page, so that new fields may be filled separately from original ones.
I tried this code:

```ruby
require 'hexapdf'

pdf = HexaPDF::Document.open('page duplicate.pdf')
page = pdf.pages[0]

(1 .. 2).each do |idx|
  # page.value.update(page.copy_inherited_values)
  copy_page = page.deep_copy

  field = copy_page[:Annots].find { |annot| annot[:Subtype] == :Widget && annot[:T].start_with?('name_') }

  field[:T] = "name_#{idx}"

  pdf.catalog[:Pages].insert_page(page.index + idx, copy_page)
end

pdf.write('page duplicate result 1.pdf')
```

The resulting PDF contains three pages, with text fields, but when I fill in text field on the first page, fields on second and third page also get filled with the same value, despite that we tried to rename those fields when copying page.

How may i fix this?

**Comment by gettalong** [comment]
So, that's currently not so easy to do. The `#deep_copy` method only deep copies direct objects, indirect objects are just referenced. Since form fields are always indirect objects, you are in effect modifying the original field.

I will have to think about how to achieve this.

**Reply by gettalong** [reply, marked as answer]
So, the following will do the trick:

~~~ ruby
require 'hexapdf'

pdf = HexaPDF::Document.open('page.duplicate.pdf')
page = pdf.pages[0]

(1..2).each do |idx|
  HexaPDF::Importer.instance_variable_get(:@map)&.delete(pdf.hash)
  copy_page = HexaPDF::Importer.for(pdf).import(page)

  field = copy_page[:Annots].find { |annot| annot[:Subtype] == :Widget && annot[:T].start_with?('name_') }
  field[:T] = "name_#{idx}"
  pdf.acro_form[:Fields] << field

  pdf.pages << copy_page
end

pdf.write('page duplicate result 1.pdf')
~~~

Since the importer doesn't currently expose a possibility to forget a certain destination document object, we need to manually delete it. I will maybe add a `HexaPDF::Importer.forget(pdf)` method to accomplish this in future HexaPDF version.

The importer does exactly what you want: A complete copy of the respective object, including the copying of indirect objects.

**Comment by ekzobrain** [comment]
It woks, thanks!
